lib/astarte_vmq_plugin.ex

```diff
@@ -272,4 +279,61 @@ defmodule Astarte.VMQ.Plugin do
       {:error, {:already_started, pid}} -> pid
     end
   end
+
+  defp authorize_registration(mountpoint, username) do
+    [realm, device_id] = String.split(username, "/")
+    {:ok, decoded_device_id} = Device.decode_device_id(device_id, allow_extended_id: true)
+
+    cond do
+      not device_exists?(realm, decoded_device_id) ->
+        {:error, :device_does_not_exist}
+
+      device_deletion_in_progress?(realm, decoded_device_id) ->
+        {:error, :device_deletion_in_progress}
+
+      true ->
+        {:ok, registration_modifiers(mountpoint, username)}
+    end
```

Comment:
while it is true that it is not immediately clear which errors might be returned, this can be easily fixed by adding a type spec to the function

Reply:
> as it would introduce two other small cluttery functions

my suggestion consisted in replacing your `device_exists?` and `device_deletion_in_progress?` functions with something returning an `:ok`/`{:error, reason}` instead of a boolean, so it wouldn't _introduce_ two new functions as much as replacing them.

> typespecs [...] should be present throughout the whole application, or else the reader would have two sources of information

while I agree it is inconsistent to have a typespec only on _some_ functions, I'd say that typespecs give an additional source of information even when used throughout the whole module or application.